### DISCUSSION TITLE
New `replaygain` opererator has no `id` property

### DISCUSSION THREAD

**Discussion by add-n2x** [savonet/liquidsoap#2537] [Q&A] [answered]
Previously I was using this to apply ReplayGain to a queue:

```ocaml
input_filesystem_0 = request.queue(id="in_filesystem_0")
input_filesystem_0 = amplify(id="in_filesystem_0", 1., override="replay_gain", input_filesystem_0)
```
After migrating to Liquidsoap 2.1 I tried the new operator:

```ocaml
input_filesystem_0 = request.queue(id="in_filesystem_0")
input_filesystem_0 = replaygain(input_filesystem_0)
```

But since the `replaygain` operator doesn't allow to pass an `id`, I cannot properly address the source after that assignment.

Do you have any tips how to solve it?

**Comment by vitoyucepi** [comment]
Hi.
You could redefine `replaygain` with all required parameters.
Reference implementation: https://github.com/savonet/liquidsoap/blob/5f69b0ffdbbfc38d9e772169d809a3c2800addb9/libs/audio.liq#L184-L203

For example this is how to add the `id` parameter.
```ruby
def replaces replaygain(~id=null(), ~ebu_r128=true, s)
  s = insert_metadata(s)
  # Normalize opus gain
  def f(m)
    def f(m)
      let (k, v) = m
      if k == "r128_track_gain" then
        v = int_of_string(v)
        v = v + 5 # normalize to -18 dB as usual replaygain instead of -23 dB
        v = lin_of_dB(float(v) / 256.)
        ("replaygain_track_gain", string_of_float(v))
      else
        (k, v)
      end
    end
    if ebu_r128 then list.map(f, m) else m end
  end
  s = metadata.map(f, s)
  amplify(id=id, override="replaygain_track_gain", 1., s)
end
```
After you redefine the function in your code, it's maintenance is up to you.

**Reply by add-n2x** [reply]
Thanks for the answer.  I'd say t's more simple to use the old operator, than redefining the new one.

**Reply by vitoyucepi** [reply]
It's not possible (I think so) to use old operator and set the `id`.
The only way is to replace `replaygain` in your code or manually patch it in library.

**Reply by add-n2x** [reply]
JFYI - the old operator is `amplify` and has an `id` property:

`amplify(id="in_filesystem_0", 1., override="replay_gain", input_filesystem_0)`

**Reply by vitoyucepi** [reply]
`amplify` has `id` arg for a long time. I can't read ocaml sources, but it was there in [*1.4.0*](https://www.liquidsoap.info/doc-1.4.0/reference.html#amplify).

Replaygain is of another kind. It's written in liquidsoap and for the first time appeared in 27f63334.

**Comment by vitoyucepi** [comment, marked as answer]
This will be fixed in
- 2.0.8
- 2.1.1

**Reply by add-n2x** [reply]
That's great! thx.
